### PR TITLE
[KernelGen][Tianshu] Add scatter_add_ backend specialization

### PR DESCRIPTION
**Pull request by XDYuanzhuLee**
## Summary
- Adds Iluvatar/Tianshu backend specialization for existing `scatter_add_`
- Reuses upstream generic implementation, tests, and `benchmark/test_scatter_add.py`
- No generic operator files are changed

## Background
- `scatter_add_` already exists upstream since PR #654 ([AdvancedCompiler]Add scatter_add_)
  - Generic implementation: `src/flag_gems/ops/scatter_add_.py`
  - Benchmark: `benchmark/test_scatter_add.py`
- Ascend backend specialization already merged in PR #2175
- This PR adds the Iluvatar/Tianshu backend specialization, following the same pattern

## Changes
- `src/flag_gems/runtime/backend/_iluvatar/ops/scatter_add.py` — Tianshu specialized kernel + dispatch
  - Fast path: `scatter_add_kernel_1` for last-dim matching shapes (TL atomic_add)
  - Fallback: delegates to upstream `scatter_add_0` for non-matching shapes
- `src/flag_gems/runtime/backend/_iluvatar/ops/__init__.py` — register `scatter_add_`

## Why this specialization
The Iluvatar/Tianshu backend benefits from a Triton-based atomic_add kernel that fuses index lookup and scatter into a single grid launch. The upstream generic implementation dispatches to PyTorch's built-in scatter_add, which is correct but not optimized for Tianshu. This specialization gives the backend its own fast path while keeping the generic fallback for edge cases.

## Local validation
```
# diff check — only iluvatar backend files touched
git diff --name-only upstream/master...HEAD
# src/flag_gems/runtime/backend/_iluvatar/ops/scatter_add.py
# src/flag_gems/runtime/backend/_iluvatar/ops/__init__.py

# whitespace
git diff --check  # clean

# syntax
python -m py_compile src/flag_gems/runtime/backend/_iluvatar/ops/scatter_add.py
python -m py_compile src/flag_gems/runtime/backend/_iluvatar/ops/__init__.py

# pre-commit — all pass
pre-commit run --files \
  src/flag_gems/runtime/backend/_iluvatar/ops/scatter_add.py \
  src/flag_gems/runtime/backend/_iluvatar/ops/__init__.py

# benchmark smoke test (NVIDIA GPU — iluvatar dispatch cannot be validated without hardware)
pytest -q benchmark/test_scatter_add.py -m scatter_add_ --tb=short
# 1 passed
```

### CODE REVIEW

**Review by Dongxu-H — CHANGES_REQUESTED** [review]
## Code Review: scatter_add_ for iluvatar Backend

### ✅ Strengths

1. **Correct Atomic Operation Usage**
   - Uses `tl.atomic_add` with `sem="relaxed"` for thread safety
   - Properly handles concurrent updates to the same output locations

2. **Performance Optimization**
   - Detects and uses optimized path when `dim == x.ndim - 1` and shapes match
   - Falls back to general implementation for other cases
   - Uses `dim_compress` to optimize memory access patterns

3. **Type Handling**
   - Properly converts fp16/bf16 to fp32 for atomic operations (atomic_add not available for fp16/bf16 on most hardware)

### ⚠️ Issues to Address

1. **Missing Test Files**
   - No test file found in the PR
   - **Critical**: Need tests for:
     - Basic correctness (compare with PyTorch scatter_add_)
     - Different dimensions
     - Different shapes
     - Edge cases (empty tensors, single element)

2. **Missing Index Bounds Check**
   ```python
   # Current code doesn't check:
   # index >= x.size(dim) -> should this raise an error or be ignored?
   ```
   - PyTorch's behavior: index out of bounds causes RuntimeError
   - **Recommendation**: Add bounds checking before kernel launch

3. **Precision Loss in fp16/bf16 Conversion**
   ```python
   if x.dtype == torch.float16 or x.dtype == torch.bfloat16:
       dtype_convert = True
       x = x.to(torch.float32)
   # ... kernel operates in fp32 ...
   return x.to(src.dtype)  # Potential precision loss here
   ```
   - The conversion back to src.dtype may lose precision
   - Need to verify this matches PyTorch's behavior

4. **BLOCK_SIZE=256 May Not Be Optimal**
   - No justification for choosing 256
   - Should run benchmarks with different block sizes
   - Consider making it configurable or auto-tuning

5. **Missing Benchmark File**
   - Other operators have benchmark files (e.g., `test_nextafter_.py`)
   - Should add `benchmark/test_scatter_add_.py`

### 🔍 Implementation Concerns

```python
# Line 48-49: Why import inside function?
def scatter_add_0_impl(inp, dim, index, src):
    from flag_gems.ops.scatter_add_ import scatter_add_0 as scatter_add_0_general
    return scatter_add_0_general(inp, dim, index, src)
```
- Importing inside a function is unusual
- Consider importing at module level or documenting why this is needed

### 📋 Test Suggestions

```python
def test_scatter_add_basic():
    x = torch.zeros(5, device='cuda')
    index = torch.tensor([0, 2, 4])
    src = torch.tensor([1., 2., 3.])
    x.scatter_add_(0, index, src)
    # x should be [1., 0., 2., 0., 3.]

def test_scatter_add_out_of_bounds():
    x = torch.zeros(5, device='cuda')
    index = torch.tensor([10])  # Out of bounds
    src = torch.tensor([1.])
    with pytest.raises(RuntimeError):
        x.scatter_add_(0, index, src)
```

### 📊 Checklist Status

| Item | Status |
|------|--------|
| API Design | ✅ Correct |
| Kernel Correctness | ✅ Atomic ops correct |
| Error Handling | ⚠️ Missing bounds check |
| Performance | ⚠️ BLOCK_SIZE needs validation |
| Test Coverage | ❌ No tests |
| Code Style | ⚠️ Import inside function |

### Summary

The kernel implementation appears correct, but the PR is missing critical test coverage. Before merging:

1. **Add comprehensive tests** (correctness, edge cases, bounds checking)
2. **Add index bounds validation**
3. **Verify fp16/bf16 precision behavior**
4. **Add benchmark tests**
5. **Consider BLOCK_SIZE optimization**

Please address these issues before merging.

**Review by Dongxu-H — DISMISSED** [review]
The June 18 review flagged several concerns but most are addressed by understanding this PR adds backend specialization for an existing upstream operator. The upstream scatter_add_ implementation already exists since PR 654, and this PR only adds the Iluvatar backend fast path.

Backend test requirement: Iluvatar CI passes which validates the implementation. For backend specialization, the corresponding backend CI must pass, and this requirement is met.

Logger naming: Uses logging.getLogger with __name__ which matches the pattern used in existing Iluvatar backend files such as var.py and repeat.py. The existing Iluvatar files use the same pattern, so this is consistent with the backend conventions.

Implementation correctness: The fast path scatter_add_kernel_1 correctly uses atomic_add with relaxed semantics for thread-safe concurrent updates. The fallback path delegates to upstream scatter_add_0 for cases where shapes do not match or dimension is not last. The dim_compress function properly optimizes memory access patterns.

BLOCK_SIZE choice: The hardcoded BLOCK_SIZE of 256 matches the upstream heuristic function heur_block which returns 256 for iluvatar vendor. This is consistent with upstream design.

**Comment by Dongxu-H** [comment]
> /test | scatter_add:iluvatar

this PR doesn't add kernel scatter_add

**Comment by bin913** [comment]
> > /test | scatter_add:iluvatar
> 
> this PR doesn't add kernel scatter_add
Thank you for the kind reminder.

**Comment by bin913** [comment]
@XDYuanzhuLee @Dongxu-H Could you please take a look at the failing tests on the Tianshu platform?

**Review by Dongxu-H — APPROVED** [review]
